here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/models/storage/ReadFeaturesProxyWrapper.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (C) 2017-2023 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.lib.core.models.storage;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ReadFeaturesProxyWrapper extends ReadFeatures {
```

Comment:
I understand the need for passing original request information down the chain to HttpStorage. That being said there are two issues I have with current `ProxyWrapper`
1) It is very REST oriented and I think `core` module should not even take `REST` under consideration (it's the REST/handler's job to comply with `core` , not the other way around)
2) I favor composition over inheritance - it is easier to deal with and allows more flexibility

I was thinking of a solution that would address both of these points above - consider following suggestion:
- We create a very simple interface called `RequestOrigin` that would be basically a read-only store for `request` metadata (it might have some `Map getProperties()` method)
- We introduce `final @Nullable RequestOrigin origin` property (with getter)  in `ReadFeatures` class 
- We create `RestEndointOrigin` class that would implement `RequestOrigin` and supply metadata about http request that this `ReadFeatures` got created from (like url & headers)
- All of the `ReadFeatures` instances created at API level would be supplied with `origin` corresponding to given endpoint
- `HttpStorage` instead of casting `ReadFeatures` to `ProxyWrapper` could just call `getOrigin().getProperties()` method and extract metadata needed for some logic (like establishing `ReadrequestType` that no longer needs to reside in `core`)
- This would mean that apart from changes in `ReadFeatures` we don't have to introduce anything new. 
	- Creating origin happens on API level
	- Handling metadata from origin happens in handler

This is just an idea though - would be awesome to hear what you think about it :)

Reply:
Actually, use of inheritance is an agreed design approach as there can be multiple cases overwriting core ReadFeatures object, as long as relevant Handlers and Storages understand the respective implementations. Yes, it is possible that this can open a bit of pandora box, but I don't see any other better option to manage the flexibility we need to support custom implementations.
While for this particular case, it may be possible to update ReadFeatures and add Origin, but I see high chances that this will only address current use-case and will most likely remain limited to this and such Http Storage based interactions only, which actually is not our dominant use-case. So, I have a bit of different view here keeping in mind other use-cases as well (for example - Validation Endorsement, has completely different extension demands of Request structure).
So, if we choose some common composition approach using ReadFeatures, then I tend to think, that it should address multiple use-cases and not just one, else it will become hybrid of composition and inheritance, which will be even more confusing to understand ;-)